File: src/main/java/io/github/starwishsama/namelessbot/objects/RandomResult.java
package io.github.starwishsama.namelessbot.objects;

import lombok.AllArgsConstructor;
import lombok.Data;

import java.text.NumberFormat;

@Data
@AllArgsConstructor
public class RandomResult {
    private long id;
    private double chance;
    private String eventName;

    private enum EventRate {
        HIGHEST("大吉"),
        HIGH("中吉"),
        NORMAL("小吉"),
        LOW("末吉"),
        LOWEST("凶"),
        NEVER("大凶");

        private String event;

        EventRate(String event) {
            this.event = event;
        }
    }

    public static String getChance(RandomResult result){
        if (result != null) {
            NumberFormat nf = NumberFormat.getPercentInstance();
            nf.setMaximumIntegerDigits(3);
            nf.setMinimumFractionDigits(2);
            double chance = result.getChance();
            return getChance(result.getEventName(), chance);
        }
        return null;
    }

    public static String getChance(String eventName, double chance){
        NumberFormat nf = NumberFormat.getPercentInstance();
        nf.setMaximumIntegerDigits(3);
        nf.setMinimumFractionDigits(2);
        String finalRate = nf.format(chance);
        if (chance > 0.8 && chance <= 1.0){
            return "结果是" + EventRate.HIGHEST.event + " (" + finalRate + "), 今天非常适合" + eventName + "!";
        } else if (chance > 0.6 && chance <= 0.8){
            return "结果是" + EventRate.HIGH.event + " (" + finalRate + "), 今天很适合" + eventName + "!";
        } else if (chance > 0.5 && chance <= 0.6){
            return "结果是" + EventRate.NORMAL.event + " (" + finalRate + "), 今天适合" + eventName + "!";
        } else if (chance > 0.3 && chance <= 0.5){
            return "结果是" + EventRate.LOW.event + " (" + finalRate + "), 今天不太适合" + eventName + "...";
        } else if (chance > 0.1 && chance <= 0.3){
            return "结果是" + EventRate.LOWEST.event + " (" + finalRate + "), 今天最好不要" + eventName + "...";
        } else if (chance <= 0.1){
            return "结果是" + EventRate.NEVER.event + " (" + finalRate + "), 千万别" + eventName + "!";
        } else {
            return "你要占卜的东西有点怪呢, 我无法占卜出结果哦.";
        }
    }
}

File: src/main/kotlin/io/github/starwishsama/comet/objects/RandomResult.kt
package io.github.starwishsama.comet.objects

import io.github.starwishsama.comet.enums.EventRate
import java.text.NumberFormat

data class RandomResult (var id: Long, var chance: Double, var eventName: String){
    companion object {
        private fun getChance(eventName: String, chance: Double): String {
            val nf = NumberFormat.getPercentInstance()
            val randomResult = "占卜结果 👇\n占卜内容:"
            nf.maximumIntegerDigits = 3
            nf.minimumFractionDigits = 2
            val finalRate = nf.format(chance)
            val rateType = when (chance) {
                in 0.8..1.0 -> EventRate.HIGHEST
                in 0.6..0.8 -> EventRate.HIGH
                in 0.5..0.6 -> EventRate.NORMAL
                in 0.3..0.5 -> EventRate.LOW
                in 0.1..0.3 -> EventRate.LOWEST
                else -> EventRate.NEVER
            }

            return "$randomResult ${eventName}\n结果: ${rateType.type} (${finalRate}%)"
        }

        fun getChance(result: RandomResult): String {
            val nf = NumberFormat.getPercentInstance()
            nf.maximumIntegerDigits = 3
            nf.minimumFractionDigits = 2
            return getChance(result.eventName, result.chance)
        }
    }
}